drivers/kubevirt/kubevirt.go

```diff
@@ -507,3 +515,350 @@ func AddPVCsToVirtualMachine(vm kubevirtv1.VirtualMachine, pvcs []*corev1.Persis
+	if err != nil {
+		return fmt.Errorf("failed to inspect volume [%s]: %w", vol.ID, err)
+	}
+	nodeIpAttachedOn := volInspect.AttachedOn
+	nodeNameAttachedOn, err := node.GetNodeByIP(nodeIpAttachedOn)
+	if err != nil {
+		return fmt.Errorf("failed to get node name by IP [%s]: %w", nodeIpAttachedOn, err)
+	}
+	log.Infof("Volume [%s] is attached on node [%s]", vol.ID, nodeNameAttachedOn.Name)
+	if nodeNameAttachedOn.Name != vmNodeName {
+		return fmt.Errorf("volume [%s] is attached on node [%s] instead of node [%s]", vol.ID, nodeNameAttachedOn.Name, vmNodeName)
+	}
+
+	isBindMounted := false
+	t := func() (interface{}, bool, error) {
+		vmPod, err := GetVirtLauncherPodForVM(virtualMachineCtx, vol)
+		if err != nil {
+			// this is expected while the live migration is running since there will be 2 VM pods
+			log.Infof("Could not get VM pod for %s for context %s: %v", vol.Name, virtualMachineCtx.App.Key, err)
+			return false, false, nil
+		}
+		log.Infof("Verifying bind mount for %s", vol)
+		diskName := ""
+		for _, vmVol := range vmPod.Spec.Volumes {
+			pvcName := volInspect.Locator.VolumeLabels["pvc"]
+			if vmVol.PersistentVolumeClaim != nil && vmVol.PersistentVolumeClaim.ClaimName == pvcName {
+				diskName = vmVol.Name
+				break
+			}
+		}
+		mountType, err := getVMDiskMountType(vmPod, vol, volumeDriver, diskName)
+		if err != nil {
+			log.Warnf("Failed to get mount type of %s for context %s: %v", vol, virtualMachineCtx.App.Key, err)
+			return false, false, nil
+		}
+		log.Infof("Mount type of %s for context %s: %s", vol.Name, virtualMachineCtx.App.Key, mountType)
+		if mountType != mountTypeBind {
+			if !initialCheck {
+				log.Warnf("Waiting for %s for context %s to switch to bind-mount from %q",
+					vol.Name, virtualMachineCtx.App.Key, mountType)
+			}
+			return false, false, nil
+		}
+		isBindMounted = true
+		return true, false, nil
+	}
+	if initialCheck {
+		// initial check is done only once
+		_, _, err := t()
+		if err != nil {
+			return err
+		}
+		if !isBindMounted {
+			return fmt.Errorf("volume [%s] is not bind mounted", vol.ID)
+		}
+	} else {
+		_, err = task.DoRetryWithTimeout(t, defaultVmMountCheckTimeout, defaultVmMountCheckRetryInterval)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// VerifyVolumeReplicaCollocation verifies if the volume replicas are collocated on the same set of nodes
+func VerifyVolumeReplicaCollocation(vol *volume.Volume, globalReplicSet []*api.ReplicaSet, driver volume.Driver) error {
+	// Check if volumes have replicas on the same set of nodes
+	volInspect, err := driver.InspectVolume(vol.ID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect volume [%s]: %w", vol.ID, err)
+	}
+
+	replicaset := volInspect.ReplicaSets
+
+	// check if the replicaset size is same as the global replicaset size
+	if len(replicaset) != len(globalReplicSet) {
+		return fmt.Errorf("replicaset count mismatch for volume [%s] and for volume [%s]", vol.ID, volInspect.Id)
+	}
```

Comment:
this check should be part of ReplicaSetsMatch function?

Reply:
Done moved it to ReplicaSetsMatch